device-api/src/topology/mod.rs

```diff
@@ -0,0 +1,349 @@
+use crate::topology::hwloc_binding::*;
+use crate::topology::LinkType::*;
+use crate::{Device, DeviceResult};
+
+mod build;
+mod hwloc;
+mod hwloc_binding;
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum LinkType {
+    // LinkTypeUnknown unknown
+    LinkTypeUnknown = 0,
+    // LinkTypeInterconnect two devices are connected across different cpus through interconnect.
+    LinkTypeInterconnect = 10,
+    // LinkTypeCPU two devices are connected under the same cpu, it may mean:
+    // devices are directly attached to the cpu pcie lane without PCIE switch.
+    // devices are attached to different PCIE switches under the same cpu.
+    LinkTypeCPU = 20,
+    // LinkTypeHostBridge two devices are connected under the same PCIE host bridge.
+    // Note that this does not guarantee devices are attached to the same PCIE switch.
+    // More switches could exist under the host bridge switch.
+    LinkTypeHostBridge = 30,
+
+    // NOTE(@bg): Score 40 and 50 is reserved for LinkTypeMultiSwitch and LinkTypeSingleSwitch.
+    // NOTE(@bg): Score 60 is reserved for LinkTypeBoard
+
+    // LinkTypeSoc two devices are on the same Soc chip.
+    LinkTypeSoc = 70,
+}
+
+impl LinkType {
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            LinkTypeUnknown => "LinkTypeUnknown",
+            LinkTypeInterconnect => "LinkTypeInterconnect",
+            LinkTypeCPU => "LinkTypeCPU",
+            LinkTypeHostBridge => "LinkTypeHostBridge",
+            LinkTypeSoc => "LinkTypeSoc",
+        }
+    }
+}
+
+pub trait HardwareTopologyHint {
+    fn get_hw_topology_hint(&self) -> String;
+}
+
+pub struct Topology {
+    hwloc_topology: Box<dyn Hwloc>,
+    topology_matrix: BTreeMap<(String, String), LinkType>,
+}
+
+impl Topology {
+    pub fn new(devices: Vec<Device>) -> DeviceResult<Topology> {
+        let mut new_topology = Self {
+            hwloc_topology: Box::new(HwlocTopology::new()),
+            topology_matrix: BTreeMap::new(),
+        };
+
+        let keys = devices.iter().map(|d| d.busname().unwrap()).collect();
+        new_topology.populate_with_keys(keys)?;
+
+        Ok(new_topology)
+    }
+
+    fn populate_with_keys(&mut self, devices: Vec<String>) -> DeviceResult<()> {
+        // Initialize hwloc topology
+        self.hwloc_topology.init_topology()?;
+
+        // Set I/O types filter
+        self.hwloc_topology
+            .set_io_types_filter(hwloc_type_filter_e_HWLOC_TYPE_FILTER_KEEP_IMPORTANT)?;
+
+        // Load the topology
+        self.hwloc_topology.load_topology()?;
+
+        match self.populate_topology_matrix(devices) {
+            Ok(matrix) => Ok(()),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn populate_topology_matrix(&mut self, devices: Vec<String>) -> DeviceResult<()> {
+        for i in 0..devices.len() {
+            for j in 0..devices.len() {
+                let dev1_bdf = devices.get(i).unwrap().clone();
+                let dev2_bdf = devices.get(j).unwrap().clone();
```

Comment:
```suggestion
        for (dev1, dev2) in iproduct!(devices, devices) {
            let dev1_bdf = dev1.busname().unwrap();
            let dev2_bdf = dev2.busname().unwrap();
```

with `use itertools::iproduct;`

Reply:
dev1, dev2 is string here

Reply:
done
25a4a133d930bf886c57bd6a0f9b7152660b9a95